lhotse/cut/multi.py

```diff
@@ -0,0 +1,1128 @@
+import logging
+from dataclasses import dataclass, field
+from functools import partial
+from math import isclose
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+import numpy as np
+import torch
+from intervaltree import IntervalTree
+
+from lhotse.audio import Recording
+from lhotse.augmentation import AugmentFn
+from lhotse.cut.base import Cut
+from lhotse.features import FeatureExtractor, Features
+from lhotse.features.io import FeaturesWriter
+from lhotse.supervision import SupervisionSegment
+from lhotse.utils import (
+    LOG_EPSILON,
+    Pathlike,
+    Seconds,
+    TimeSpan,
+    add_durations,
+    compute_num_frames,
+    compute_num_samples,
+    fastcopy,
+    ifnone,
+    measure_overlap,
+    overlaps,
+    overspans,
+    perturb_num_samples,
+    rich_exception_info,
+    uuid4,
+)
+
+
+@dataclass
+class MultiCut(Cut):
+    """
+    :class:`~lhotse.cut.MultiCut` is a :class:`~lhotse.cut.Cut` that is analogous to the MonoCut.
+    While MonoCut represents a single channel of a recording, MultiCut represents multiple channels
+    that share the same supervision. It is intended to be used to store, for example, segments
+    of a microphone array recording.
+
+    By definition, a MultiCut has the same attributes as a MonoCut. The key difference is that
+    the Recording object has multiple channels, and the Supervision object is shared across
+    multiple channels. The channels that the MultiCut pertains to is defined by the channels
+    in the Supervision object, and these are the same as the ``channel`` attribute of the
+    MultiCut. These channels may be a subset of the channels in the Recording object. For
+    example, if the Recording object has 4 channels (0, 1, 2, 3), we may construct a MonoCut
+    object with a supervision for channel 0, and a MultiCut object with a supervision for
+    channels 1 and 2.
```

Comment:
I think that MultiCut channels should be interpreted separately from Supervision channels. E.g.:

3-channel MultiCut that consists of
- 4-channel Recording (one channel is not used)
- a 2-channel Supervision (e.g. stereo mics for conversation in XY setting)
- a 1-channel Supervision (e.g. from a different mic)

Does that make sense? It seems (too?) flexible but maybe it's a good opportunity to account for these possibilities if you're doing serious revisions to the code :)

Reply:
Yeah that makes sense. I was coming in more from a utility perspective instead of a representation perspective. However, one aspect we might want to think is how would such a broader definition of MultiCut interact with things like mixing.

Reply:
Perhaps this is what you are proposing?
```
    While MonoCut represents a single channel of a recording, MultiCut represents multi-channel
    recordings where supervisions may or may not be shared across channels. 
    It is intended to be used to store, for example, segments of a microphone array recording.
    The following diagrams illustrate some examples for MultiCut usage:

    >>> 2-channel telephone recording with 2 supervisions, one for each channel (e.g., Switchboard):


                  ╔══════════════════════════════  MultiCut  ═════════════════╗
                  ║ ┌──────────────────────────┐                              ║
     Channel 1  ──╬─│   Hello this is John.    │──────────────────────────────╬────────
                  ║ └──────────────────────────┘                              ║
                  ║                               ┌──────────────────────────┐║
     Channel 2  ──╬───────────────────────────────│ Hey, John. How are you?  │╠────────
                  ║                               └──────────────────────────┘║
                  ╚═══════════════════════════════════════════════════════════╝

    >>> Multi-array multi-microphone recording with shared supervisions (e.g., CHiME-6),
    along with close-talk microphones (A and B are distant arrays, C is close-talk):

               ╔═══════════════════════════════════════════════════════════════════════════╗
               ║ ┌───────────────────┐                         ┌───────────────────┐       ║
       A-1   ──╬─┤                   ├─────────────────────────┤                   ├───────╬─
               ║ │ What did you do?  │                         │I cleaned my room. │       ║
       A-2   ──╬─┤                   ├─────────────────────────┤                   ├───────╬─
               ║ └───────────────────┘  ┌───────────────────┐  └───────────────────┘       ║
       B-1   ──╬────────────────────────┤Yeah, we were going├──────────────────────────────╬─
               ║                        │   to the mall.    │                              ║
       B-2   ──╬────────────────────────┤                   ├──────────────────────────────╬─
               ║                        └───────────────────┘        ┌───────────────────┐ ║
        C    ──╬─────────────────────────────────────────────────────┤      Right.       ├─╬─
               ║                                                     └───────────────────┘ ║
               ╚════════════════════════════════  MultiCut  ═══════════════════════════════╝
```

Reply:
Now I have to learn how to make these diagrams from you 🤯 yes, this is what I had in mind.